sw/device/silicon_creator/lib/cert/dice_cwt.c

```diff
@@ -103,33 +137,177 @@ const sc_keymgr_ecc_key_t kDiceKeyCdi1 = {
 //     -2 : bstr,                   ; X coordinate, big-endian
 //     -3 : bstr                    ; Y coordinate, big-endian
 // }
+static rom_error_t cose_key_ec_p256_build(ecdsa_p256_public_key_t *uds_pubkey,
+                                          uint8_t *cose_key_buffer,
+                                          size_t *cose_key_buffer_size) {
+  struct CborOut kCborOutHandle;
+  struct CborOut *pCborOut = &kCborOutHandle;
+
+  RETURN_IF_ERROR(
+      cbor_write_out_init(pCborOut, cose_key_buffer, *cose_key_buffer_size));
+  RETURN_IF_ERROR(cbor_map_init(pCborOut, kCoseKeyMapCount));
+  RETURN_IF_ERROR(cbor_write_pair_uint_uint(pCborOut, CWT_COSE_KEY_LABEL_TYPE,
+                                            CWT_COSE_KEY_TYPE_EC2));
+  RETURN_IF_ERROR(cbor_write_pair_uint_int(pCborOut, CWT_COSE_KEY_LABEL_ALG,
+                                           CWT_COSE_KEY_ALG_ECDSA256));
+  RETURN_IF_ERROR(cbor_write_pair_int_uint(pCborOut, CWT_COSE_KEY_LABEL_CRUVE,
+                                           CWT_COSE_EC2_CRUVE_P256));
+  RETURN_IF_ERROR(cbor_write_pair_int_bytes(pCborOut, CWT_COSE_KEY_LABEL_EC2_X,
+                                            (uint8_t *)uds_pubkey->x,
+                                            sizeof(uds_pubkey->x)));
+  RETURN_IF_ERROR(cbor_write_pair_int_bytes(pCborOut, CWT_COSE_KEY_LABEL_EC2_Y,
+                                            (uint8_t *)uds_pubkey->y,
+                                            sizeof(uds_pubkey->y)));
+  *cose_key_buffer_size = CborOutSize(pCborOut);
+
+  return kErrorOk;
+}
+
 rom_error_t dice_uds_tbs_cert_build(
     hmac_digest_t *otp_creator_sw_cfg_measurement,
     hmac_digest_t *otp_owner_sw_cfg_measurement,
     hmac_digest_t *otp_rot_creator_auth_codesign_measurement,
     hmac_digest_t *otp_rot_creator_auth_state_measurement,
     cert_key_id_pair_t *key_ids, ecdsa_p256_public_key_t *uds_pubkey,
     uint8_t *tbs_cert, size_t *tbs_cert_size) {
+  HARDENED_RETURN_IF_ERROR(
+      cose_key_ec_p256_build(uds_pubkey, tbs_cert, tbs_cert_size));
+  return kErrorOk;
+}
+
+// protected:
+// a1    # map(1)
+//    01 #   unsigned(1)
+//    26 #   negative(-7)
+const uint8_t cbor_map_algo_es256[] = {0xa1, 0x01, 0x26};
+
+// a0    # map(0)
+const uint8_t cbor_map_0 = 0xa0;
+
+// Pre-calculated SHA256 of "0xa0"
+const uint8_t sha256_of_cbor_map_0[] = {
+    0xc1, 0x9a, 0x79, 0x7f, 0xa1, 0xfd, 0x59, 0x0c, 0xd2, 0xe5, 0xb4,
+    0x2d, 0x1c, 0xf5, 0xf2, 0x46, 0xe2, 0x9b, 0x91, 0x68, 0x4e, 0x2f,
+    0x87, 0x40, 0x4b, 0x81, 0xdc, 0x34, 0x5c, 0x7a, 0x56, 0xa0};
+
+static ecdsa_p256_signature_t curr_tbs_signature = {.r = {0}, .s = {0}};
+
+/**
+ * Helper function to convert an attestation certificate signature from little
+ * to big endian.
+ */
+static void curr_tbs_signature_le_to_be_convert(ecdsa_p256_signature_t *sig) {
```

Comment:
A similar function exists in https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/cert/dice.c;drc=40107c126540a065de2de6c55c92a3f9f4ca5c49;l=122 ; lets move it to a util lib and share between dice.c and dice_cwt.c